### DISCUSSION TITLE
THEIA_HOSTS issue

### DISCUSSION THREAD

**Discussion by vishaakm** [eclipse-theia/theia#8844] [Q&A] [answered]
- `THEIA_HOSTS`
  - A comma-separated list of hosts expected to resolve to the current application.
    - e.g: `theia.app.com,some.other.domain:3000`
  - The port number is important if your application is not hosted on either `80` or `443`.
  - If possible, you should set this environment variable:
    - When not set, Theia will allow any origin to access the WebSocket services.
    - When set, Theia will only allow the origins defined in this environment variable.

I followed the above comment and didnt set the `THEIA_HOSTS` env 
but the websocket requests are still being blocked 

Need Input on this

**Comment by azatsarynnyy** [comment]
cc @marechal-p

**Comment by paul-marechal** [comment]
@vishaakm Do you have more info? Is the app not starting anymore? Are you connecting to the websockets from a third-party app?

**Reply by vishaakm** [reply]
The app starts perfectly fine also usable using the IP directly

But when I use a https reverse proxy hostname websocket gets blocked and gives 403 error

**Reply by paul-marechal** [reply]
Is this reverse proxy configuration working with Theia 1.8.0?

**Reply by vishaakm** [reply]
yes it was working fine

used this
```
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $http_connection;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP  $remote_addr;
        proxy_set_header X-Forwarded-For $remote_addr;
        add_header Service-Worker-Allowed /;
```

**Reply by paul-marechal** [reply]
Can you try adding  `proxy_set_header Origin $http_origin`?

**Reply by paul-marechal** [reply]
The following setup worked for me:

### NGINX configuration

```nginx
events {}

http {
    server {
        listen 4000;
        location /a/b/ {
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $http_connection;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP  $remote_addr;
            proxy_set_header X-Forwarded-For $remote_addr;
            add_header Service-Worker-Allowed /;
            proxy_pass http://127.0.0.1:3000/;
        }
    }
}
```

### Running NGINX with docker locally

```sh
docker run --name nginx-theia-1.8.1 --network="host" -v $(pwd)/nginx.conf:/etc/nginx/nginx.conf:ro -d nginx:stable-alpine
```

### Adding an entry to my `/etc/hosts` file

```hosts
127.0.0.1 www.patate.local
```

### Accessing the app

URL: http://www.patate.local:4000/a/b

---

@vishaakm is your setup much different?

**Comment by paul-marechal** [comment, marked as answer]
After more investigation, @vishaakm's issue is most likely caused by setting one of `THEIA_WEBVIEW_ENDPOINT` and/or `THEIA_MINI_BROWSER_HOST_PATTERN` to `{{hostname}}`.

The reason is that the patch included in 1.8.1 will reject requests with origins from webviews and the mini-browser, despite not setting `THEIA_HOSTS`. I did not have in mind setups where the origin for those services were the same as the main application.

This issue will be fixed in 1.9.0, which should come soon.

**Reply by paul-marechal** [reply]
@vishaakm please confirm that you set those variables and removing them fixes the issue.

**Reply by vishaakm** [reply]
@marechal-p Removing `THEIA_WEBVIEW_EXTERNAL_ENDPOINT` made it to work, but now the webviews are not being displayed :(

**Reply by paul-marechal** [reply]
See https://github.com/eclipse-theia/theia/issues/8852
